### PR TITLE
Send Docker image with CI build

### PR DESCRIPTION
**Pull request by LTA-Thinking**
## Description
Send a docker image to an Azure container repository upon successful CI build.

## Related issues
#779 #840 

## Testing
This PR is currently being tested by sending an image to a test container repository upon PR build. This will be changed to be upon CI build and be sent to an official repository.

### CODE REVIEW

File: build/.vsts-pr.yml [hunk 1]
```diff
@@ -4,6 +4,8 @@
 name: pr$(system.pullRequest.pullRequestNumber)-$(Date:yyyyMMdd)$(Rev:-r)
 variables:
   buildConfiguration: 'Release'
+  azureSubscriptionEndpoint: 'docker-test'
+  azureContainerRegistry: 'ossdockertest.azurecr.io'
```
Comment (deepak-ms) [inline build/.vsts-pr.yml]: Should we use HealthPlatformRegistry for this? It has anonymous pull enabled (i.e. public access).

File: build/.vsts-pr.yml [hunk 2]
```diff
@@ -18,3 +20,38 @@ jobs:
...
+
+  - task: DockerCompose@0
+    displayName: Build FHIR Server Image
+    inputs:
+      action: Build services
+      dockerComposeFile: samples/docker/docker-compose.yaml 
+      projectName: fhirserverforazure
+      qualifyImageNames: true
+      additionalImageTags: |
+        pr$(system.pullRequest.pullRequestNumber)
```
Comment (deepak-ms) [inline build/.vsts-pr.yml]: Does this build image from PR branch or master?

File: build/.vsts-pr.yml [hunk 3]
```diff
@@ -18,3 +20,38 @@ jobs:
...
+    inputs:
+      action: Push services
+      azureSubscriptionEndpoint: $(azureSubscriptionEndpoint)
+      azureContainerRegistry: $(azureContainerRegistry)
+      dockerComposeFile: docker-compose.yml
+      projectName: fhirserverforazure
+      qualifyImageNames: true
+      additionalImageTags: |
+        pr$(system.pullRequest.pullRequestNumber)
+        $(Build.BuildId)
```
Comment (deepak-ms) [inline build/.vsts-pr.yml]: You might want to define these as variables. Also do we need 'latest' tag too?

File: build/.vsts-pr.yml (hunk 1 above)
Comment (LTA-Thinking) [reply build/.vsts-pr.yml]: That's a good idea for the final repo. Right now I'm trying to get it to work, so I think a test account is best.

File: build/.vsts-pr.yml (hunk 2 above)
Comment (LTA-Thinking) [reply build/.vsts-pr.yml]: For now, the PR branch. When it builds off of CI it will want to pull from master.

**Comment by LTA-Thinking** [comment]
Steps to run the FHIR server's most recent version from a Docker image:
1. Open a terminal window.
2. Login to the Azure Container Registry 'ossdockertest' (this will change before finishing this PR, most likely to the HealthPlatformRegistry suggested above) (https://docs.microsoft.com/en-us/azure/container-registry/container-registry-get-started-docker-cli)
3. Set the enviornment variable SAPASSWORD to what you want the SQL access password to be. Be sure to follow the SQL password complexity rules: https://docs.microsoft.com/en-us/sql/relational-databases/security/password-policy?view=sql-server-ver15#password-complexity.
4. Copy & save a local version of the docker-compose file from the release directory of the fhir-server project.
5. Run the command: docker-compose up -d.
6. After giving the container a few minutes to start up it should be accessable at http://localhost/metadata.

**Comment by LTA-Thinking** [comment]
To do this without the docker-compose file a network will need to be made between the SQL server image and the FHIR server image. Replace the <sql-sa-password> below with your chosen SQL connection password, following the complexity rules found here: https://docs.microsoft.com/en-us/sql/relational-databases/security/password-policy?view=sql-server-ver15#password-complexity.
1. Open a terminal window
2. Login to the Azure Container Registry 'ossdockertest' (https://docs.microsoft.com/en-us/azure/container-registry/container-registry-get-started-docker-cli)
3. Run the command: docker network create fhir_network
4. Run the command: docker run --net fhir_network --name fhir_sql -e SA_PASSWORD=<sql-sa-password> -e ACCEPT_EULA="Y" -d mcr.microsoft.com/mssql/server
5. Give the SQL server ~1 minute to start up
6. Run the command: docker run --net fhir_network -e FhirServer__Security__Enabled="false" -e SqlServer__ConnectionString="Server=tcp:fhir_sql,1433;Initial Catalog=FHIR;Persist Security Info=False;User ID=sa;Password=<sql-sa-password>;MultipleActiveResultSets=False;Connection Timeout=30;" -e SqlServer__AllowDatabaseCreation="true" -e SqlServer__Initialize="true" -e DataStore="SqlServer" -p 80:80 -d ossdockertest.azurecr.io/fhirserverforazure_fhir-api azure-fhir-api
7. After giving the container a minute to start up it should be accessible at http://localhost/metadata.

File: samples/docker/README.md [hunk 4]
```diff
@@ -4,6 +4,48 @@
 
 The following instructions detail how to build and run the FHIR Server in Docker on Linux.
 
+## Use CI image
+
+If it is not desirable to clone this repository and build locally an image of the most recent CI build is available from the Azure Container Repository HealthPlatformRegistry. 
+
+Using docker-compose this image can be started with the following steps:
+1. Open a terminal window.
+2. Login to the Azure Container Registry 'healthplatformregistry' (https://docs.microsoft.com/en-us/azure/container-registry/container-registry-get-started-docker-cli)
```
Comment (deepak-ms) [inline samples/docker/README.md]: Is login step still required (with anonymous pull enabled registry)?

File: samples/docker/README.md [hunk 5]
```diff
@@ -4,6 +4,48 @@
 
 The following instructions detail how to build and run the FHIR Server in Docker on Linux.
 
+## Use CI image
+
```
Comment (deepak-ms) [inline samples/docker/README.md]: Do we plan to provide "Deploy to Azure" button using the containers?

File: samples/docker/README.md [hunk 6]
```diff
@@ -4,6 +4,48 @@
...
+```bash
+docker run --net fhir_network --name fhir_sql -e SA_PASSWORD=<SA_PASSWORD> -e ACCEPT_EULA="Y" -d mcr.microsoft.com/mssql/server
+```
+
+5. Give the SQL server a minute to start up.
+6. Run the command: 
+
+```bash
+docker run --net fhir_network -e FhirServer__Security__Enabled="false" -e SqlServer__ConnectionString="Server=tcp:fhir_sql,1433;Initial Catalog=FHIR;Persist Security Info=False;User ID=sa;Password=<SA_PASSWORD>;MultipleActiveResultSets=False;Connection Timeout=30;" -e SqlServer__AllowDatabaseCreation="true" -e SqlServer__Initialize="true" -e DataStore="SqlServer" -p 80:80 -d healthplatformregistry.azurecr.io/fhirserverR4_fhir-api azure-fhir-api
```
Comment (deepak-ms) [inline samples/docker/README.md]: does it provide persistence? if no, should we add steps here?

File: samples/docker/README.md (hunk 5 above)
Comment (LTA-Thinking) [reply samples/docker/README.md]: We already have one here: https://github.com/microsoft/fhir-server/blob/master/docs/DefaultDeployment.md#deploying-the-fhir-server-template
The next story (#72981) will update the deployment script to use the container.

File: samples/docker/README.md (hunk 6 above)
Comment (LTA-Thinking) [reply samples/docker/README.md]: Added note that persistence is not provided with this deployment method.

File: samples/docker/README.md (hunk 4 above)
Comment (LTA-Thinking) [reply samples/docker/README.md]: Good catch, removed

File: build/.vsts-ci.yml [hunk 7]
```diff
@@ -17,4 +19,71 @@ jobs:
   steps:
   - template: build.yml
     parameters:
-      packageArtifacts: false
\ No newline at end of file
+      packageArtifacts: false
+      
+- job: Docker
+  pool:
+    vmImage: 'ubuntu-16.04'
```
Comment (brendankowitz) [inline build/.vsts-ci.yml]: should we switch to `ubuntu-latest`?

File: build/.vsts-ci.yml [hunk 8]
```diff
@@ -17,4 +19,71 @@ jobs:
...
+
+  - task: DockerCompose@0
+    displayName: Push FHIR R4 Server Image
+    inputs:
+      action: Push services
+      azureSubscriptionEndpoint: $(azureSubscriptionEndpoint)
+      azureContainerRegistry: $(azureContainerRegistry)
+      dockerComposeFile: build/docker/docker-compose.yaml
+      projectName: r4
+      additionalImageTags: $(SourceBranchName)
```
Comment (brendankowitz) [inline build/.vsts-ci.yml]: Looks like this could be a template step

File: build/docker/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,25 @@
+FROM mcr.microsoft.com/dotnet/core/sdk:3.1.100-bionic AS build
```
Comment (brendankowitz) [inline build/docker/Dockerfile]: Should we delete the /samples/docker/Dockerfile now we have this one?

File: build/docker-build-push.yml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+- name: tag
+  type: string
+
+steps:
+- task: DockerCompose@0
+  displayName: 'Build FHIR ${{parameters.version}} Server Image'
+  inputs:
+    action: Build services
+    azureSubscriptionEndpoint: 'docker-build'
+    azureContainerRegistry: 'healthplatformregistry.azurecr.io'
```
Comment (brendankowitz) [inline build/docker-build-push.yml]: prefer the variables if it doesn't add any extra complexity :P 
Or, is `azureContainerRegistry` required for the build if the next step is push?

Comment (LTA-Thinking) [reply build/docker-build-push.yml]: Refactored the variables into a template

File: build/docker/Dockerfile (hunk 9 above)
Comment (LTA-Thinking) [reply build/docker/Dockerfile]: Good catch. Deleted

File: build/.vsts-ci.yml (hunk 8 above)
Comment (LTA-Thinking) [reply build/.vsts-ci.yml]: Refactored into two templates

File: build/.vsts-ci.yml (hunk 7 above)
Comment (LTA-Thinking) [reply build/.vsts-ci.yml]: Replaced

File: build/.vsts-pr.yml (hunk 3 above)
Comment (LTA-Thinking) [reply build/.vsts-pr.yml]: The latest tag is provided automatically. Tags changed to input parameters.

**Pull request merged**